Objects/stringlib/find_max_char.h

```diff
@@ -19,24 +19,50 @@
 Py_LOCAL_INLINE(Py_UCS4)
 STRINGLIB(find_max_char)(const STRINGLIB_CHAR *begin, const STRINGLIB_CHAR *end)
 {
-    const unsigned char *p = (const unsigned char *) begin;
-
-    while (p < end) {
-        if (_Py_IS_ALIGNED(p, ALIGNOF_SIZE_T)) {
-            /* Help register allocation */
-            const unsigned char *_p = p;
-            while (_p + SIZEOF_SIZE_T <= end) {
-                size_t value = *(const size_t *) _p;
-                if (value & UCS1_ASCII_CHAR_MASK)
-                    return 255;
-                _p += SIZEOF_SIZE_T;
-            }
-            p = _p;
-            if (p == end)
-                break;
-        }
-        if (*p++ & 0x80)
+    const unsigned char *restrict p = (const unsigned char *) begin;
+    const unsigned char *_end = (const unsigned char *)end;
+    const unsigned char *size_t_end = _end - SIZEOF_SIZE_T;
+    const unsigned char *unrolled_end = _end - (4 * SIZEOF_SIZE_T - 1);
+    while (p < unrolled_end) {
+        /* Test chunks of 32 as more granularity limits compiler optimization */
+        const size_t *restrict _p = (const size_t *)p;
+        size_t value0;
+        size_t value1;
+        size_t value2;
+        size_t value3;
+        /* Will be optimized to simple loads for architectures that support
+           unaligned loads. */
+        memcpy(&value0, _p + 0, SIZEOF_SIZE_T);
+        memcpy(&value1, _p + 1, SIZEOF_SIZE_T);
+        memcpy(&value2, _p + 2, SIZEOF_SIZE_T);
+        memcpy(&value3, _p + 3, SIZEOF_SIZE_T);
+        size_t value = value0 | value1 | value2 | value3;
+        if (value & UCS1_ASCII_CHAR_MASK) {
             return 255;
+        }
+        p += (4 * SIZEOF_SIZE_T);
+    }
+    size_t accumulator = 0;
+    while (p < size_t_end) {
+        size_t value;
+        memcpy(&value, p, SIZEOF_SIZE_T);
+        accumulator |= value;
+        p += SIZEOF_SIZE_T;
+    }
+    if (size_t_end >= (const unsigned char*)begin) {
```

Comment:
I don't understand this test. Why testing **begin** here?

Reply:
I should add a comment here. 

Because size_t_end is the end - SIZEOF_SIZE_T. That means we can do an unaligned load from size_t_end and load 8 characters at once for the bitwise OR. This is quicker than checking a remainder of 1-7 bytes. 

But if the string is smaller than 8 characters this means it will load characters that are not part of the string. That is very bad.

Reply:
> I should add a comment here.

Yes, please.

Reply:
Done